### PR TITLE
feat: add support for working with locale

### PR DESCRIPTION
**Pull request by mhevery**
# What is it?

- [ ] Feature / enhancement
- [ ] Bug
- [ ] Docs / tests

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# Use cases and why

- 1. One use case
- 2. Another use case

# Checklist:

- [ ] My code follows the [developer guidelines of this project](../CONTRIBUTING.md)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

### CODE REVIEW

**Comment by manucorporat** [comment]
This PR introduces a new API, which implementation i am not sure it's entirely correct since use methods can only be used in  rendering context, what's this for?

File: packages/qwik/src/core/render/container.ts [hunk 1]
```diff
@@ -79,6 +79,17 @@ export const createContainerState = (containerEl: Element) => {
   return containerState;
 };
 
+const envDataFromContainer = (containerEl: Element): Record<string, any> => {
```
Comment (manucorporat) [inline packages/qwik/src/core/render/container.ts:82]: envData is not container attributes, we are mixing semantics, not sure i like this, we are telling developers that envData is server only (for now), but it comes from `envData`, this breaks that semantic

File: packages/qwik/src/core/api.md [hunk 2]
```diff
@@ -268,6 +268,9 @@ export const inlinedQrl: <T>(symbol: T, symbolName: string, lexicalScopeCapture?
 // @internal (undocumented)
 export const inlinedQrlDEV: <T = any>(symbol: T, symbolName: string, opts: QRLDev, lexicalScopeCapture?: any[]) => QRL<T>;
 
+// @alpha
+export const isInUseContext: () => boolean;
```
Comment (manucorporat) [inline packages/qwik/src/core/api.md:272]: i would make it `@internal` is we really need it, the fact that `_context` is defined does not mean that all use methods will work. Only useLexicalScope() works in all contexts

**Review by manucorporat — CHANGES_REQUESTED** [review]

File: packages/qwik-city/middleware/request-handler/page-handler.ts [hunk 3]
```diff
@@ -130,14 +137,21 @@ function getPrefetchBundleNames(result: RenderResult, routeBundleNames: string[]
   return bundleNames;
 }
 
-export function getQwikCityEnvData(userResponse: UserResponseContext): {
+export function getQwikCityEnvData(
+  requestHeaders: Record<string, string>,
+  userResponse: UserResponseContext,
+  mode: QwikCityMode
+): {
```
Comment (wmertens) [inline packages/qwik-city/middleware/request-handler/page-handler.ts:144]: Would it not be better to switch this to an object argument to allow adding parameters in the future?

File: packages/qwik-city/middleware/request-handler/page-handler.ts [hunk 4]
```diff
@@ -21,6 +26,8 @@ export function pageHandler<T = any>(
   const { status, headers } = userResponse;
   const { response } = requestCtx;
   const isPageData = userResponse.type === 'pagedata';
+  const requestHeaders: Record<string, string> = {};
+  requestCtx.request.headers.forEach((value, key) => (requestHeaders[key] = value));
```
Comment (wmertens) [inline packages/qwik-city/middleware/request-handler/page-handler.ts:30]: We have something similar in our React codebase, but we allow registering functions that extra environment data from a request object.

This way, more than headers can be used to set the environment data, it avoids having to copy all headers, and it allows running the render without a request object, like when you're rendering a PDF.

**Review by wmertens — COMMENTED** [review]
- how can one set the first part of the URL to the locale?
- how can one change the locale in the browser without losing state, e.g. while filling in a form?

File: packages/qwik/src/core/render/container.ts (hunk 1 above)
Comment (wmertens) [reply packages/qwik/src/core/render/container.ts:82]: in our React codebase, we have plugins that can extract server data during render (via `<Provider>` mosty), and then populate a global state object that gets serialized to the client, which gets read by the Providers on the client.
This has proven very useful, and I think this `envData` is similar to that.

**Comment by DustinJSilk** [comment]
This is great, it also solves a [PR](https://github.com/BuilderIO/qwik/pull/1605) I sent to get access to request headers.

**Pull request merged**